File: studio/web-app/src/app/api-bak/ui-model.ts
// direct import to avoid require all common in ui
// we depend on mylife-home-ui only for this
// export * from 'mylife-home-ui/dist/shared/model';
// FIXME: regenerate it, no cross imports
export * from '../../../../../ui/web-app/src/app/api/model';

// export * from 'mylife-home-ui/dist/src/model/definition';

import { Window, DefaultWindow } from '../../../../../ui/web-app/src/app/api/model';

export interface Definition {
  readonly resources: DefinitionResource[];
  readonly styles: DefinitionStyle[];
  readonly windows: Window[];
  readonly defaultWindow: DefaultWindow;
}

export interface DefinitionResource {
  readonly id: string;
  readonly mime: string;
  readonly data: string;
}

export interface DefinitionStyle {
  readonly id: string;
  readonly properties: object;
}

File: studio/web-api/src/ui_model.rs
use serde::{Deserialize, Serialize};
use ts_rs::TS;

use crate::{
    model::{Action, ActionComponent, ActionWindow, Control, ControlDisplay, ControlDisplayMapItem, ControlText, ControlTextContextItem, DefaultWindow, Resource, Style, Window},
    register_ts,
};

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export_to = "ui-model.ts")]
pub struct Definition {
    pub resources: Vec<DefinitionResource>,
    pub styles: Vec<DefinitionStyle>,
    pub windows: Vec<Window>,
    pub default_window: DefaultWindow,
}

register_ts!(Definition);

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export_to = "ui-model.ts")]
pub struct DefinitionResource {
    pub id: String,
    pub mime: String,
    pub data: String,
}

register_ts!(DefinitionResource);

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export_to = "ui-model.ts")]
pub struct DefinitionStyle {
    pub id: String,
    #[ts(type = "object")]
    pub properties: serde_json::Value,
}

register_ts!(DefinitionStyle);

// Re-export the same model names most of the UI app relies on.
register_ts!(Resource);
register_ts!(Style);
register_ts!(DefaultWindow);
register_ts!(Window);
register_ts!(Control);
register_ts!(ControlDisplay);
register_ts!(ControlDisplayMapItem);
register_ts!(ControlText);
register_ts!(ControlTextContextItem);
register_ts!(Action);
register_ts!(ActionComponent);
register_ts!(ActionWindow);
